http/src/main/scala/org/http4s/blaze/http/http2/HeaderDecoder.scala

```diff
@@ -0,0 +1,91 @@
+package org.http4s.blaze.http.http2
+
+import java.nio.ByteBuffer
+import java.nio.charset.StandardCharsets.US_ASCII
+
+import org.http4s.blaze.http.http2.Http2Exception._
+import org.http4s.blaze.http.http2.Http2Settings.DefaultSettings
+import org.http4s.blaze.util.BufferTools
+import com.twitter.hpack.{Decoder, HeaderListener}
+
+import scala.collection.immutable.VectorBuilder
+
+/** Decoder of HEADERS frame payloads into a collection of key-value
+  * pairs using a HPACK decoder.
+  *
+  * @param maxHeaderListSize maximum allowed size of the header block.
+  *
+  * "The value is based on the uncompressed size of header fields,
+  *  including the length of the name and value in octets plus an
+  *  overhead of 32 octets for each header field."
+  * https://tools.ietf.org/html/rfc7540#section-6.5.2
+  * @param maxTableSize maximum compression table to maintain
+  */
+final class HeaderDecoder(
+  maxHeaderListSize: Int,
+  discardOverflowHeaders: Boolean,
+  val maxTableSize: Int
+) {
+  require(maxTableSize >= DefaultSettings.HEADER_TABLE_SIZE)
+
+  private[this] val acc = new VectorBuilder[(String, String)]
+  private[this] var leftovers: ByteBuffer = null
+  private[this] var headerBlockSize = 0
+
+  private[this] val decoder = new Decoder(maxHeaderListSize, maxTableSize)
+  private[this] val listener = new HeaderListener {
+    override def addHeader(name: Array[Byte], value: Array[Byte], sensitive: Boolean): Unit = {
+      headerBlockSize += 32 + name.length + value.length
+      if (!discardOverflowHeaders || !headerListSizeOverflow) {
+        acc += new String(name, US_ASCII) -> new String(value, US_ASCII)
+        ()
+      }
+    }
+  }
+
+  /** Set the HEADER_TABLE_SIZE parameter */
+  def setMaxHeaderTableSize(max: Int): Unit = decoder.setMaxHeaderTableSize(max)
+
+  /** Returns the header collection and clears the builder */
+  def result(): Seq[(String,String)] = {
```

Comment:
leftovers is kept for a subsequent run, right?

Reply:
It should always be empty after we've gotten to the end of the header block and want to call result. Maybe this could be made better by changing the name to `finish` and enforcing that it has been called after `decode(.., endHeaders = true)`.

Reply:
Okay. My concern was that it wasn't being cleared when the rest of the state was.

Reply:
Yeah, and it's a good call just to clear it: who knows if the hpack decoder could have left some stay padding laying around, so good eye.